aptos-move/framework/aptos-framework/sources/configs/Stake.move

```diff
@@ -278,25 +485,47 @@ module AptosFramework::Stake {
     /// This function shouldn't abort.
-    fun update_stake_pool(addr: address) acquires StakePool {
-        let stake_pool = borrow_global_mut<StakePool>(addr);
-        distribute_reward( &mut stake_pool.active);
-        distribute_reward( &mut stake_pool.pending_inactive);
-        // move pending_active to active
-        IterableTable::append(&mut stake_pool.active, &mut stake_pool.pending_active);
-        // move pending_inactive to inactive
-        IterableTable::append(&mut stake_pool.inactive, &mut stake_pool.pending_inactive);
-    }
-
-    /// Mint the reward and add to the delegation based on some formula
-    fun distribute_reward(v: &mut IterableTable<address, Delegation>) {
-        let key = IterableTable::head_key(v);
-        while (Option::is_some(&key)) {
-            let (delegation, _, next) = IterableTable::borrow_iter_mut(v, *Option::borrow(&key));
-            let reward = Coin::zero<TestCoin>(); // mint some coins based on delegation, timestamp, maybe also total stakes
-            Coin::merge<TestCoin>(&mut delegation.rewards, reward);
-            key = next;
+    fun update_stake_pool(
+        validator_perf: &ValidatorPerformance,
+        pool_address: address,
+    ) acquires StakePool, TestCoinCapabilities {
+        let stake_pool = borrow_global_mut<StakePool>(pool_address);
+        let num_missed_votes = *Vector::borrow(&validator_perf.missed_votes, stake_pool.validator_index);
+        let num_blocks = validator_perf.num_blocks;
+        let num_successful_votes = num_blocks - num_missed_votes;
+
+        let current_time = Timestamp::now_seconds();
+        // Remaining lockup time needs to be at least one so the rewards calculation still works.
+        let remaining_lockup_time = 1;
+        if (current_time < stake_pool.locked_until_secs) {
+            remaining_lockup_time = stake_pool.locked_until_secs - current_time;
         };
+
+        distribute_reward(&mut stake_pool.active, num_blocks, num_successful_votes, remaining_lockup_time);
+        distribute_reward(&mut stake_pool.pending_inactive, num_blocks, num_successful_votes, remaining_lockup_time);
+
+        // move pending_active to active
+        Coin::merge<TestCoin>(&mut stake_pool.active, Coin::extract_all<TestCoin>(&mut stake_pool.pending_active));
+        Coin::merge<TestCoin>(&mut stake_pool.inactive, Coin::extract_all<TestCoin>(&mut stake_pool.pending_inactive));
+    }
+
+    /// Distribute reward corresponding to `stake` and `num_successful_votes` and `remaining_Lockup_time`.
+    fun distribute_reward(
+        stake: &mut Coin<TestCoin>,
+        num_blocks: u64,
+        num_successful_votes: u64,
+        remaining_lockup_time: u64,
+    ) acquires TestCoinCapabilities {
+        // Validators receive rewards based on their performance (number of successful votes) and how long is their
+        // remaining lockup time.
+        let base_rewards = Coin::value<TestCoin>(stake) * REWARDS_RATE_PERCENTAGE / 100;
```

Comment:
I imagine this is going to be more granular than 2 digits can provide.

Reply:
This all depends on the decimals of TestCoin. As long as it has enough decimals to cover the per-epoch rewards for the min stake, it should be good.